hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMRecoverLeaseRequest.java

```diff
@@ -0,0 +1,281 @@
+    Iterator<Path> pathComponents = Paths.get(keyName).iterator();
+    long parentID = OMFileRequest.getParentID(volumeId, bucketId,
+        pathComponents, keyName, omMetadataManager,
+        "Cannot recover file : " + keyName
+            + " as parent directory doesn't exist");
+    String fileName = OzoneFSUtils.getFileName(keyName);
+    String dbFileKey = omMetadataManager.getOzonePathKey(volumeId, bucketId,
+        parentID, fileName);
+
+    OmKeyInfo keyInfo = getKey(dbFileKey);
+    if (keyInfo == null) {
+      throw new OMException("Key:" + keyName + " not found", KEY_NOT_FOUND);
+    }
+    final String clientId = keyInfo.getMetadata().get(
+        OzoneConsts.HSYNC_CLIENT_ID);
+    if (clientId == null) {
+      LOG.warn("Key:" + keyName + " is closed");
+    }
+    String openKeyEntryName = getOpenFileEntryName(volumeId, bucketId,
+        parentID, fileName);
+    if (openKeyEntryName != null) {
+      checkFileState(keyInfo, openKeyEntryName);
+      commitKey(dbFileKey, keyInfo, ozoneManager, transactionLogIndex);
+      removeOpenKey(openKeyEntryName, transactionLogIndex);
+    }
+
+    return openKeyEntryName;
+  }
+
+  private OmKeyInfo getKey(String dbOzoneKey) throws IOException {
+    return omMetadataManager.getKeyTable(getBucketLayout()).get(dbOzoneKey);
+  }
+
+  private String getOpenFileEntryName(long volumeId, long bucketId,
+      long parentObjectId, String fileName) throws IOException {
+    String openFileEntryName = null;
+    String dbOpenKeyPrefix =
+        omMetadataManager.getOpenFileNamePrefix(volumeId, bucketId,
+            parentObjectId, fileName);
+    try (TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
+        iter = omMetadataManager.getOpenKeyTable(getBucketLayout())
+        .iterator(dbOpenKeyPrefix)) {
+
+      while (iter.hasNext()) {
+        if (openFileEntryName != null) {
+          throw new IOException("Found more than two keys in the" +
+            " open key table for file" + fileName);
+        }
+        openFileEntryName = iter.next().getKey();
+      }
+    }
+    /*if (openFileEntryName == null) {
+      throw new IOException("Unable to find the corresponding key in " +
+        "the open key table for file " + fileName);
+    }*/
+    return openFileEntryName;
+  }
+
+  private void checkFileState(OmKeyInfo keyInfo, String openKeyEntryName)
+      throws IOException {
+
+    // if file is closed, do nothing and return right away.
+    if (openKeyEntryName.isEmpty()) {
+      return; // ("File is closed");
+    }
+    // if file is open, no sync, fail right away.
+    if (keyInfo == null) {
+      throw new IOException("file is open but not yet sync'ed");
+    }
+  }
+
+  private void commitKey(String dbOzoneKey, OmKeyInfo omKeyInfo,
+      OzoneManager ozoneManager, long transactionLogIndex) throws IOException {
+    omKeyInfo.setModificationTime(Time.now());
+    omKeyInfo.setUpdateID(transactionLogIndex, ozoneManager.isRatisEnabled());
```

Comment:
Please check impact of multi-part key present in openKey table for FSO bucket type, its commit have different behavior and can not commit in that case.

Reply:
IMO i'm inclined to not support lease recovery and hsync for MPK. Hsync and lease recovery is a behavior tailor for HDFS applications and MPK was not support in HDFS. But let me check if we need to add guardrails or tests to verify.